core/src/main/java/org/apache/accumulo/fate/Fate.java

```diff
@@ -71,24 +83,51 @@ public void run() {
         try {
           tid = store.reserve();
           TStatus status = store.getStatus(tid);
+          if (status == FAILED) {
+            runnerLog.info(
+                "FATE txid " + Long.toHexString(tid) + " likely cancelled before it started.");
+            return;
+          }
           Repo<T> op = store.top(tid);
-          if (status == TStatus.FAILED_IN_PROGRESS) {
+          if (status == FAILED_IN_PROGRESS) {
             processFailed(tid, op);
           } else {
             Repo<T> prevOp = null;
+            boolean ieSeen = false;
             try {
-              deferTime = op.isReady(tid, environment);
-
-              if (deferTime == 0) {
-                prevOp = op;
-                if (status == TStatus.SUBMITTED) {
-                  store.setStatus(tid, TStatus.IN_PROGRESS);
+              try {
+                synchronized (RUNNING_TRANSACTIONS) {
+                  RUNNING_TRANSACTIONS.put(tid, Thread.currentThread());
                 }
-                op = op.call(tid, environment);
-              } else
-                continue;
-
+                deferTime = op.isReady(tid, environment);
+                if (deferTime == 0) {
+                  prevOp = op;
+                  if (status == SUBMITTED) {
+                    store.setStatus(tid, IN_PROGRESS);
+                  }
+                  op = op.call(tid, environment);
+                } else
+                  continue;
+              } catch (InterruptedException ie) {
+                ieSeen = true;
+                throw ie;
+              } finally {
+                synchronized (RUNNING_TRANSACTIONS) {
+                  RUNNING_TRANSACTIONS.remove(tid);
+                }
+                // It's possible that cancel was called between
+                // op.call and the synchronized block above. Do a last
+                // check to see if that's the case.
+                if (!ieSeen && Thread.currentThread().isInterrupted()) {
+                  throw new InterruptedException("FaTE thread was interrupted");
+                }
```

Comment:
If there is any exception on the try{} block then we probably do not want to throw anything in the finally block as it will obscure the original exception.  Generalizing the code as follows may avoid this problem.

```suggestion
              } catch (Exception e) {
                eSeen = true;
                throw e;
              } finally {
                synchronized (RUNNING_TRANSACTIONS) {
                  RUNNING_TRANSACTIONS.remove(tid);
                }
                // It's possible that cancel was called between
                // op.call and the synchronized block above. Do a last
                // check to see if that's the case.
                if (!eSeen && Thread.currentThread().isInterrupted()) {
                  throw new InterruptedException("FaTE thread was interrupted");
                }
```

Reply:
resolved in [111820f](https://github.com/apache/accumulo/pull/2467/commits/111820f8e93b0163e9bc69194a8442dfc3522e4b)